boxoffice/models/discount_policy.py

```diff
@@ -109,6 +110,63 @@ def make_bulk(cls, discount_code_base, **kwargs):
         """
         return cls(discount_type=DISCOUNT_TYPE.COUPON, discount_code_base=discount_code_base, secret=buid(), **kwargs)
 
+    __roles__ = {
+        'dp_owner': {
+            'write': {},
+            'read': {'id', 'name', 'title', 'is_automatic', 'item_quantity_min', 'percentage', 'is_price_based', 'discount_code_base', 'line_items_count'}
+        }
+    }
+
+    @classmethod
+    def get_from_item(cls, item, qty, coupon_codes=[]):
+        """
+        Returns a list of (discount_policy, discount_coupon) tuples
+        applicable for an item, given the quantity of line items and coupons if any.
+        """
+        automatic_discounts = item.discount_policies.filter(cls.discount_type == DISCOUNT_TYPE.AUTOMATIC,
+            cls.item_quantity_min <= qty).all()
+        policies = [(discount, None) for discount in automatic_discounts]
+        if not coupon_codes:
+            return policies
+        else:
+            coupon_policies = item.discount_policies.filter(cls.discount_type == DISCOUNT_TYPE.COUPON).all()
+            coupon_policy_ids = [cp.id for cp in coupon_policies]
+            for coupon_code in coupon_codes:
+                coupons = []
+                if cls.is_signed_code_format(coupon_code):
+                    policy = cls.get_from_signed_code(coupon_code)
+                    if policy and policy.id in coupon_policy_ids:
+                        coupon = DiscountCoupon.query.filter_by(discount_policy=policy, code=coupon_code).one_or_none()
+                        if not coupon:
+                            coupon = DiscountCoupon(
+                                discount_policy=policy,
+                                code=coupon_code,
+                                usage_limit=policy.bulk_coupon_usage_limit,
+                                used_count=0)
+                            db.session.add(coupon)
+                        coupons.append(coupon)
+                else:
+                    coupons = DiscountCoupon.query.filter(
+                        DiscountCoupon.discount_policy_id.in_(coupon_policy_ids),
+                        DiscountCoupon.code == coupon_code).all()
+
+                for coupon in coupons:
+                    if coupon.usage_limit > coupon.used_count:
+                        policies.append((coupon.discount_policy, coupon))
+        return policies
+
+    @property
+    def line_items_count(self):
+        from ..models import LineItem, LINE_ITEM_STATUS
+
+        return LineItem.query.filter(LineItem.discount_policy == self, LineItem.status == LINE_ITEM_STATUS.CONFIRMED).count()
```

Comment:
Probably for a future PR, but if we added `line_items` as a relationship in this model, instead of the other way around, these queries will get much simpler. ` return self.line_items.filter(LineItem.status == LINE_ITEM_STATUS.CONFIRMED).count()`.

Reply:
Fixed.